---
title: "CTD_timeseries"
author: "Matomo Niwano"
date: "3/17/2021"
output: html_document
---

## DADA2 AMPLICON ANALYSIS - CTD time-series
## Merge runs: Chimeras & Taxonomy

## Version of programs and R packages used 

- **R** = 4.0
- **R Studio** = 1.3.1093
- **dada2** = 1.14.1

```{r}

# This script: combine sequence tables of individual MiSeq runs
# Chimera removal and taxonomy assignment

# Load packages
library(dada2)
library(ShortRead)
library(ggplot2)
library(gridExtra)

# Change working directory
setwd("/isibhv/projects/FRAMdata/maniwano/WaterCol_CTD_1519/finalData")

# Read data from each run
st_J3FNT <- readRDS(
  "/isibhv/projects/FRAMdata/maniwano/WaterCol_CTD_1519/J3FNT/code/J3FNT.rds")
st_JBKFT <- readRDS(
  "/isibhv/projects/FRAMdata/maniwano/WaterCol_CTD_1519/JBKFT/code/JBKFT.rds")
st_JBL8W <- readRDS(
  "/isibhv/projects/FRAMdata/maniwano/WaterCol_CTD_1519/JBL8W/code/JBL8W.rds")

# Merge sequence tables
seqtab <- mergeSequenceTables(
  st_J3FNT, st_JBKFT, st_JBL8W, repeats="error")

# Remove chimeras 
# 13070 bimeras in 19656 input sequences -- NOTE: this is quite high, but acceptable (see lines 60-61) 
seqtab.nochim <- removeBimeraDenovo(
  seqtab, method = "consensus", 
  multithread = 6, verbose = T)
  
# stats
dim(seqtab.nochim)  # 6586 sequences in 206 samples
summary(rowSums(seqtab.nochim)/rowSums(seqtab))
# Min = 0.8643
# 1st Qu. = 0.9742
# Median = 0.9802
# Mean = 0.9805
# 3rd Qu. = 0.9897
# Max = 0.9959
sum(seqtab.nochim)/sum(seqtab) # 0.9802457
# so the high number of chimeras corresponded to many low-abundance reads
# hence, the overall community profile is not really affected

# Determine read lengths/size range of amplicons
table(rep(nchar(colnames(seqtab.nochim)), 
          colSums(seqtab.nochim)))

# Remove singletons and "junk" sequences
# "c" adjusted to size range of amplicons 
seqtab.nochim2 <- seqtab.nochim[, nchar(
  colnames(seqtab.nochim)) %in% c(353:413) & 
    colSums(seqtab.nochim) > 1]

# stats
dim(seqtab.nochim2) # 6293 sequences
summary(rowSums(seqtab.nochim2))
# Min. 1st Qu.  Median  Mean   3rd Qu.  Max. 
# 69    76696  109164  125213  150053  519575 
summary(rowSums(seqtab.nochim2)/rowSums(seqtab.nochim))
# Min.    1st Qu.  Median  Mean   3rd Qu.  Max. 
# 0.9966  0.9994  0.9999  0.9996  1.0000  1.0000 
```

## TAXONOMY -- Silva v138 ##

```{r}

# download from https://zenodo.org/record/3731176
tax_silva138 <- assignTaxonomy(
  seqtab.nochim2, 
  "silva_nr_v138_train_set.fa.gz", 
  tryRC = TRUE,
  multithread = 10)

# Archaea - 215  Bacteria - 5654 Eukaryota - 87
summary(tax_silva138)

# Remove NA on phylum level
table(tax_silva138[, 1])   
sum(is.na(tax_silva138[, 2]))   # 305
tax.good <- tax_silva138[!is.na(tax_silva138[, 2]),]
seqtab.nochim2.good <- seqtab.nochim2[, rownames(tax.good)]
summary(rowSums(seqtab.nochim2.good))
 # Min. 1st Qu.  Median    Mean   3rd Qu.    Max. 
 # 69   76156    108588   124634  149666    518010 

# Format tables
seqtab.nochim2.print <- t(seqtab.nochim2.good)
tax.print <- tax.good
all.equal(rownames(seqtab.nochim2.print), 
          rownames(tax.print)) #needs to be TRUE -- confirmed
rownames(seqtab.nochim2.print) <- paste(
  "asv", 1:ncol(seqtab.nochim2.good), sep = "")
rownames(tax.print) <- rownames(seqtab.nochim2.print)

# Export
write.table(
  seqtab.nochim2.print, "seqtab.txt", sep="\t", quote=F)
write.table(
  tax.print, "tax.txt", sep="\t", quote=F)
uniquesToFasta(
  seqtab.nochim2.good, "uniques.fasta")

# save image
save.image("WaterCol_CTD_1519.Rdata")